// paperclip-utils/lib/node/source-watcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaperclipResourceWatcher = exports.ChangeKind = void 0;
var chokidar = require("chokidar");
var path = require("path");
var url = require("url");
var events_1 = require("events");
var utils_1 = require("../core/utils");
var __1 = require("..");
var ChangeKind;
(function (ChangeKind) {
    ChangeKind[ChangeKind["Removed"] = 0] = "Removed";
    ChangeKind[ChangeKind["Added"] = 1] = "Added";
    ChangeKind[ChangeKind["Changed"] = 2] = "Changed";
})(ChangeKind = exports.ChangeKind || (exports.ChangeKind = {}));
var CHOKIDAR_EVENT_MAP = {
    add: ChangeKind.Added,
    unlink: ChangeKind.Removed,
    change: ChangeKind.Changed
};
var PaperclipResourceWatcher = /** @class */ (function () {
    function PaperclipResourceWatcher(_srcDir, cwd) {
        this._srcDir = _srcDir;
        this.cwd = cwd;
        this._em = new events_1.EventEmitter();
        this._init();
    }
    PaperclipResourceWatcher.prototype.onChange = function (listener) {
        var _this = this;
        this._em.on("change", listener);
        return function () { return _this._em.off("change", listener); };
    };
    PaperclipResourceWatcher.prototype.dispose = function () {
        this._watcher.close();
    };
    PaperclipResourceWatcher.prototype._init = function () {
        var _this = this;
        var watcher = (this._watcher = chokidar.watch((0, utils_1.paperclipSourceGlobPattern)(this._srcDir), { cwd: this.cwd, ignoreInitial: true }));
        watcher.on("all", function (eventName, relativePath) {
            if ((0, __1.isGeneratedPaperclipFile)(relativePath)) {
                return;
            }
            var filePath = relativePath.charAt(0) !== "/"
                ? path.join(_this.cwd, relativePath)
                : relativePath;
            var changeKind = CHOKIDAR_EVENT_MAP[eventName];
            if (changeKind) {
                _this._em.emit("change", changeKind, url.pathToFileURL(filePath).href);
            }
        });
    };
    return PaperclipResourceWatcher;
}());
exports.PaperclipResourceWatcher = PaperclipResourceWatcher;
