// paperclip-interim/lib/test/basic-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var utils_1 = require("paperclip/lib/test/utils");
var __1 = require("..");
var utils_2 = require("./utils");
describe(__filename + "#", function () {
    [
        [
            "can convert a simple AST to a module",
            {
                "/src/entry.pc": "<div export component as=\"Test\">\n        </div>"
            },
            {},
            { html: "<style></style> <div></div>" }
        ],
        [
            "Can embed assets",
            {
                "/src/entry.pc": "\n          <style>\n            @font-face {\n              src: url(/c.svg);\n            }\n            .a {\n              background-image: url(/d.svg);\n            }\n            .b {\n              background-image: url(/b.svg);\n            }\n          </style>\n          <img export component as=\"Test\" src=\"/a.svg\" />\n          <img export component as=\"Test2\" src=\"/b.svg\" />\n          <div export component as=\"A\">\n            {<img src=\"/e.svg\" />}\n          </div>\n        ",
                "/a.svg": "embedded-svg",
                "/b.svg": "not-embedded-svg",
                "/c.svg": "embedded",
                "/d.svg": "embed",
                "/e.svg": "embed2"
            },
            {
                embedAssetMaxSize: "embedded-svg".length,
                useAssetHashNames: false
            },
            {
                html: "<style>@font-face { src:url(data:image/svg+xml;base64,ZW1iZWRkZWQ=); } [class]._a61d499e_a { background-image:url(data:image/svg+xml;base64,ZW1iZWQ=); } [class]._a61d499e_b { background-image:url(/b.svg); }</style> <img src=data:image/svg+xml;base64,ZW1iZWRkZWQtc3Zn></img> <img src=/b.svg></img> <div><img src=data:image/svg+xml;base64,ZW1iZWQy></img></div>"
            }
        ],
        [
            "Renames assets to assetOutDir",
            {
                "/src/entry.pc": "\n          <style>\n            @font-face {\n              src: url(./c.svg);\n            }\n            .a {\n              background-image: url(./d.svg);\n            }\n            .b {\n              background-image: url(./b.svg);\n            }\n          </style>\n          <img export component as=\"Test\" src=\"/src/a.svg\" />\n          <img export component as=\"Test2\" src=\"/src/b.svg\" />\n          <div export component as=\"A\">\n            {<img src=\"/src/e.svg\" />}\n          </div>\n        ",
                "/src/a.svg": "embedded-svg",
                "/src/b.svg": "not-embedded-svg",
                "/src/c.svg": "embedded",
                "/src/d.svg": "embed",
                "/src/e.svg": "embed2"
            },
            {
                assetOutDir: "./lib",
                srcDir: "/src",
                useAssetHashNames: false
            },
            {
                html: "<style>@font-face { src:url(../lib/c.svg); } [class]._a61d499e_a { background-image:url(../lib/d.svg); } [class]._a61d499e_b { background-image:url(../lib/b.svg); }</style> <img src=../lib/a.svg></img> <img src=../lib/b.svg></img> <div><img src=../lib/e.svg></img></div>"
            }
        ],
        [
            "Can embed assets and emit files that exceed max size",
            {
                "/src/entry.pc": "\n          <style>\n            @font-face {\n              src: url(/c.svg);\n            }\n            .a {\n              background-image: url(/d.svg);\n            }\n            .b {\n              background-image: url(/b.svg);\n            }\n          </style>\n          <img export component as=\"Test\" src=\"/a.svg\" />\n          <img export component as=\"Test2\" src=\"/b.svg\" />\n          <div export component as=\"A\">\n            {<img src=\"/e.svg\" />}\n          </div>\n        ",
                "/a.svg": "embedded-svg",
                "/b.svg": "not-embedded-svg",
                "/c.svg": "embedded",
                "/d.svg": "embed",
                "/e.svg": "embed2"
            },
            {
                embedAssetMaxSize: "embedded-svg".length,
                assetOutDir: "./lib",
                srcDir: "/src",
                outDir: "/lib"
            },
            {
                html: "<style>@font-face { src:url(data:image/svg+xml;base64,ZW1iZWRkZWQ=); } [class]._a61d499e_a { background-image:url(data:image/svg+xml;base64,ZW1iZWQ=); } [class]._a61d499e_b { background-image:url(./86098dd56eddbba6fbc9cd7f03ccd8c1.svg); }</style> <img src=data:image/svg+xml;base64,ZW1iZWRkZWQtc3Zn></img> <img src=./86098dd56eddbba6fbc9cd7f03ccd8c1.svg></img> <div><img src=data:image/svg+xml;base64,ZW1iZWQy></img></div>"
            }
        ],
        [
            "Can define an asset prefix",
            {
                "/src/entry.pc": "\n          <style>\n            @font-face {\n              src: url(/c.svg);\n            }\n            .a {\n              background-image: url(/d.svg);\n            }\n            .b {\n              background-image: url(/b.svg);\n            }\n          </style>\n          <img export component as=\"Test\" src=\"/a.svg\" />\n          <img export component as=\"Test2\" src=\"/b.svg\" />\n          <div export component as=\"A\">\n            {<img src=\"/e.svg\" />}\n          </div>\n        ",
                "/a.svg": "embedded-svg",
                "/b.svg": "not-embedded-svg",
                "/c.svg": "embedded",
                "/d.svg": "embed",
                "/e.svg": "embed2"
            },
            {
                embedAssetMaxSize: "embedded-svg".length,
                assetOutDir: "./lib",
                srcDir: "/src",
                outDir: "/lib",
                assetPrefix: "http://localhost:3000/"
            },
            {
                html: "<style>@font-face { src:url(data:image/svg+xml;base64,ZW1iZWRkZWQ=); } [class]._a61d499e_a { background-image:url(data:image/svg+xml;base64,ZW1iZWQ=); } [class]._a61d499e_b { background-image:url(http://localhost:3000/lib/86098dd56eddbba6fbc9cd7f03ccd8c1.svg); }</style> <img src=data:image/svg+xml;base64,ZW1iZWRkZWQtc3Zn></img> <img src=http://localhost:3000/lib/86098dd56eddbba6fbc9cd7f03ccd8c1.svg></img> <div><img src=data:image/svg+xml;base64,ZW1iZWQy></img></div>"
            }
        ],
        [
            "Includes PC imports",
            {
                "/src/entry.pc": "\n          <import src=\"/a.pc\" as=\"test\" />\n        ",
                "/a.pc": "<div></div>"
            },
            {
                embedAssetMaxSize: "embedded-svg".length,
                assetOutDir: "./lib",
                srcDir: "/src",
                outDir: "/lib",
                assetPrefix: "http://localhost:3000/"
            },
            {
                html: "<import src=/a.pc as=test  /><style></style>"
            }
        ],
        [
            "Omits css imports",
            {
                "/src/entry.pc": "\n          <import src=\"/a.pc\" as=\"test\" />\n          <import src=\"/a.css\" inject-styles />\n        ",
                "/a.pc": "<div></div>",
                "/a.css": ".a { }"
            },
            {
                embedAssetMaxSize: "embedded-svg".length,
                assetOutDir: "./lib",
                srcDir: "/src",
                outDir: "/lib",
                assetPrefix: "http://localhost:3000/"
            },
            {
                html: "<import src=/a.pc as=test  /><style></style>"
            }
        ],
        [
            "Includes CSS imports if importAssetsAsModules is true",
            {
                "/src/entry.pc": "\n          <import src=\"/a.pc\" as=\"test\" />\n          <import src=\"/a.css\" inject-styles />\n        ",
                "/a.pc": "<div></div>",
                "/a.css": ".a { }"
            },
            {
                importAssetsAsModules: true
            },
            {
                html: "<import src=/a.pc as=test  /><import src=/a.scoped.css as=undefined  /><style></style>"
            }
        ]
    ].forEach(function (_a) {
        var title = _a[0], graph = _a[1], _b = _a[2], embedAssetMaxSize = _b.embedAssetMaxSize, assetOutDir = _b.assetOutDir, srcDir = _b.srcDir, outDir = _b.outDir, assetPrefix = _b.assetPrefix, useAssetHashNames = _b.useAssetHashNames, importAssetsAsModules = _b.importAssetsAsModules, expectedOutput = _a[3];
        it(title, function () {
            var engine = (0, utils_1.createMockEngine)(graph);
            var interim = new __1.InterimCompiler(engine, {
                cwd: "/",
                config: {
                    srcDir: srcDir
                },
                targetOptions: {
                    outDir: outDir,
                    importAssetsAsModules: importAssetsAsModules,
                    assetPrefix: assetPrefix,
                    embedAssetMaxSize: embedAssetMaxSize,
                    assetOutDir: assetOutDir,
                    useAssetHashNames: useAssetHashNames
                },
                io: {
                    readFile: function (filePath) {
                        return Buffer.from(graph[filePath]);
                    },
                    getFileSize: function (filePath) {
                        return graph[filePath].length;
                    }
                }
            });
            var module = interim.parseFile("/src/entry.pc");
            (0, chai_1.expect)((0, utils_2.stringifyInterimModule)(module)).to.eql(expectedOutput.html);
        });
    });
});
