// paperclip-utils/lib/core/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.engineDelegateChanged = exports.previewContent = exports.loadedDataEmitted = exports.loadedDataRequested = exports.astRequested = exports.astEmitted = exports.actionCreator = exports.BasicPaperclipActionType = void 0;
var BasicPaperclipActionType;
(function (BasicPaperclipActionType) {
    BasicPaperclipActionType["AST_REQUESTED"] = "AST_REQUESTED";
    BasicPaperclipActionType["LOADED_DATA_REQUESTED"] = "LOADED_DATA_REQUESTED";
    BasicPaperclipActionType["LOADED_DATA_EMITTED"] = "LOADED_DATA_EMITTED";
    BasicPaperclipActionType["AST_EMITTED"] = "AST_EMITTED";
    BasicPaperclipActionType["ENGINE_DELEGATE_CHANGED"] = "ENGINE_DELEGATE_CHANGED";
    BasicPaperclipActionType["PREVIEW_CONTENT"] = "PREVIEW_CONTENT";
})(BasicPaperclipActionType = exports.BasicPaperclipActionType || (exports.BasicPaperclipActionType = {}));
var actionCreator = function (type) { return function (payload) { return ({
    type: type,
    payload: payload
}); }; };
exports.actionCreator = actionCreator;
exports.astEmitted = (0, exports.actionCreator)(BasicPaperclipActionType.AST_EMITTED);
exports.astRequested = (0, exports.actionCreator)(BasicPaperclipActionType.AST_REQUESTED);
exports.loadedDataRequested = (0, exports.actionCreator)(BasicPaperclipActionType.LOADED_DATA_REQUESTED);
exports.loadedDataEmitted = (0, exports.actionCreator)(BasicPaperclipActionType.LOADED_DATA_EMITTED);
exports.previewContent = (0, exports.actionCreator)(BasicPaperclipActionType.PREVIEW_CONTENT);
exports.engineDelegateChanged = (0, exports.actionCreator)(BasicPaperclipActionType.ENGINE_DELEGATE_CHANGED);
