// paperclip-interim/lib/translate/script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.translateScript = void 0;
var paperclip_utils_1 = require("paperclip-utils");
var script_1 = require("../state/script");
var html_1 = require("./html");
var translateScript = function (script, context) {
    switch (script.scriptKind) {
        case paperclip_utils_1.ScriptExpressionKind.Group: {
            return {
                kind: script_1.InterimScriptExpressionKind.Group,
                inner: (0, exports.translateScript)(script.expression, context),
                range: script.range
            };
        }
        case paperclip_utils_1.ScriptExpressionKind.Node: {
            return {
                kind: script_1.InterimScriptExpressionKind.Element,
                element: (0, html_1.translateElement)(script, context),
                range: script.range
            };
        }
        case paperclip_utils_1.ScriptExpressionKind.Not: {
            return {
                kind: script_1.InterimScriptExpressionKind.Not,
                expression: (0, exports.translateScript)(script.expression, context),
                range: script.range
            };
        }
        case paperclip_utils_1.ScriptExpressionKind.Number: {
            return {
                kind: script_1.InterimScriptExpressionKind.Number,
                value: String(script.value),
                range: script.range
            };
        }
        case paperclip_utils_1.ScriptExpressionKind.Reference: {
            return {
                kind: script_1.InterimScriptExpressionKind.Reference,
                name: script.path[0].name,
                optional: script.path[0].optional,
                range: script.range
            };
        }
        case paperclip_utils_1.ScriptExpressionKind.Boolean: {
            return {
                kind: script_1.InterimScriptExpressionKind.Boolean,
                value: script.value,
                range: script.range
            };
        }
        case paperclip_utils_1.ScriptExpressionKind.String: {
            return {
                kind: script_1.InterimScriptExpressionKind.String,
                value: script.value,
                range: script.range
            };
        }
        case paperclip_utils_1.ScriptExpressionKind.Conjunction: {
            return {
                kind: script_1.InterimScriptExpressionKind.Conjunction,
                operator: script.operator,
                left: (0, exports.translateScript)(script.left, context),
                right: (0, exports.translateScript)(script.right, context),
                range: script.range
            };
        }
    }
};
exports.translateScript = translateScript;
