// paperclip-utils/lib/css/style-inspection.js
"use strict";
// 1:1 with Rust
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectorInfoKind = exports.SelectorScopeKind = void 0;
var SelectorScopeKind;
(function (SelectorScopeKind) {
    SelectorScopeKind["Element"] = "Element";
    SelectorScopeKind["Document"] = "Document";
})(SelectorScopeKind = exports.SelectorScopeKind || (exports.SelectorScopeKind = {}));
var SelectorInfoKind;
(function (SelectorInfoKind) {
    SelectorInfoKind["List"] = "list";
    SelectorInfoKind["All"] = "All";
    SelectorInfoKind["Element"] = "Element";
    SelectorInfoKind["PseudoElement"] = "PseudoElement";
    SelectorInfoKind["PsuedoParamElement"] = "PseudoParamElement";
    SelectorInfoKind["Attribute"] = "Attribute";
    SelectorInfoKind["Not"] = "Not";
    SelectorInfoKind["Id"] = "Id";
    SelectorInfoKind["Class"] = "Class";
    SelectorInfoKind["Combo"] = "Combo";
    SelectorInfoKind["Child"] = "Child";
    SelectorInfoKind["Descendent"] = "Descendent";
    SelectorInfoKind["Adjacent"] = "Adjacent";
    SelectorInfoKind["Sibling"] = "Sibling";
})(SelectorInfoKind = exports.SelectorInfoKind || (exports.SelectorInfoKind = {}));
