// paperclip-interim/lib/state/html.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.traverseInterimExpression = exports.InterimAttributeValuePartKind = exports.InterimNodeKind = void 0;
var script_1 = require("./script");
var InterimNodeKind;
(function (InterimNodeKind) {
    InterimNodeKind["Element"] = "Element";
    InterimNodeKind["Component"] = "Component";
    InterimNodeKind["Text"] = "Text";
    InterimNodeKind["Slot"] = "Slot";
})(InterimNodeKind = exports.InterimNodeKind || (exports.InterimNodeKind = {}));
var InterimAttributeValuePartKind;
(function (InterimAttributeValuePartKind) {
    InterimAttributeValuePartKind["Static"] = "Static";
    InterimAttributeValuePartKind["Dynamic"] = "Dynamic";
    InterimAttributeValuePartKind["Shorthand"] = "Shorthand";
})(InterimAttributeValuePartKind = exports.InterimAttributeValuePartKind || (exports.InterimAttributeValuePartKind = {}));
var traverseInterimExpression = function (expr, each) {
    if (each(expr) === false) {
        return false;
    }
    if (expr.kind === InterimNodeKind.Component ||
        expr.kind === InterimNodeKind.Element) {
        expr.children.forEach(function (child) {
            (0, exports.traverseInterimExpression)(child, each);
        });
    }
    else if (expr.kind === InterimNodeKind.Slot) {
        (0, exports.traverseInterimExpression)(expr.script, each);
    }
    else if (expr.kind === script_1.InterimScriptExpressionKind.Element) {
        (0, exports.traverseInterimExpression)(expr.element, each);
    }
    else if (expr.kind === script_1.InterimScriptExpressionKind.Group) {
        (0, exports.traverseInterimExpression)(expr.inner, each);
    }
    else if (expr.kind === script_1.InterimScriptExpressionKind.Conjunction) {
        (0, exports.traverseInterimExpression)(expr.left, each) &&
            (0, exports.traverseInterimExpression)(expr.right, each);
    }
};
exports.traverseInterimExpression = traverseInterimExpression;
