// paperclip-interim/lib/state/script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterimConjunctionOperator = exports.InterimScriptExpressionKind = void 0;
var InterimScriptExpressionKind;
(function (InterimScriptExpressionKind) {
    InterimScriptExpressionKind["Reference"] = "Reference";
    InterimScriptExpressionKind["Conjunction"] = "Conjunction";
    InterimScriptExpressionKind["Number"] = "Number";
    InterimScriptExpressionKind["Boolean"] = "Boolean";
    InterimScriptExpressionKind["Group"] = "Group";
    InterimScriptExpressionKind["Not"] = "Not";
    InterimScriptExpressionKind["String"] = "String";
    InterimScriptExpressionKind["Element"] = "ScriptElement";
})(InterimScriptExpressionKind = exports.InterimScriptExpressionKind || (exports.InterimScriptExpressionKind = {}));
var InterimConjunctionOperator;
(function (InterimConjunctionOperator) {
    InterimConjunctionOperator["Or"] = "Or";
    InterimConjunctionOperator["And"] = "And";
})(InterimConjunctionOperator = exports.InterimConjunctionOperator || (exports.InterimConjunctionOperator = {}));
